pam/gdm/extension_test.go

```diff
@@ -0,0 +1,241 @@
+		},
+
+		"The private string extension unsupported if not advertised": {
+			checkExtensions:     []string{PamExtensionCustomJSON},
+			supportedExtensions: nil,
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			// These tests can't be parallel since they act on env variables
+			t.Cleanup(pam_test.MaybeDoLeakCheck)
+
+			AdvertisePamExtensions(tc.advertisedExtensions)
+			t.Cleanup(func() { AdvertisePamExtensions(nil) })
+
+			for _, ext := range tc.checkExtensions {
+				shouldSupport := slices.Contains(tc.supportedExtensions, ext)
+				require.Equal(t, shouldSupport, IsPamExtensionSupported(ext))
+			}
+		})
+	}
+}
+
+func TestGdmJSONProto(t *testing.T) {
+	t.Parallel()
+	t.Cleanup(pam_test.MaybeDoLeakCheck)
+
+	testCases := map[string]struct {
+		value []byte
+	}{
+		"With null data": {
+			value: []byte("null"),
+		},
+
+		"With single int": {
+			value: []byte("55"),
+		},
+
+		"With single float": {
+			value: []byte("5.5"),
+		},
+
+		"With single string": {
+			value: []byte(`"hello"`),
+		},
+
+		"With single boolean": {
+			value: []byte("true"),
+		},
+
+		"With empty object": {
+			value: []byte("{}"),
+		},
+
+		"With complex object": {
+			value: []byte(`{"type":"pollResponse","pollResponse":` +
+				`[{"type":"brokerSelected","brokerSelected":{"brokerId":"a broker"}},` +
+				`{"type":"authModeSelected","authModeSelected":{"authModeId":"auth mode"}}]}`),
+		},
+	}
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Cleanup(pam_test.MaybeDoLeakCheck)
+
+			req, err := NewBinaryJSONProtoRequest(tc.value)
+			require.NoError(t, err)
+			t.Cleanup(req.Release)
+			require.NotNil(t, req)
+			require.NotNil(t, req.Pointer())
+			require.Equal(t, pam.BinaryPrompt, req.Style())
+
+			decoded, err := decodeJSONProtoMessage(req.Pointer())
+			require.NoError(t, err)
+			require.Equalf(t, tc.value, decoded, "JSON mismatch '%s' vs '%s'",
+				string(tc.value), string(decoded))
+		})
+	}
+}
+
+func TestGdmJSONProtoRequestErrors(t *testing.T) {
+	t.Parallel()
+	t.Cleanup(pam_test.MaybeDoLeakCheck)
+
+	testCases := map[string]struct {
+		value []byte
+	}{
+		"With empty data": {
+			value: []byte{},
+		},
+
+		"With null data": {
+			value: nil,
+		},
+
+		"With single char": {
+			value: []byte("m"),
+		},
+
+		"With lorem ipsum string data": {
+			value: []byte(`
+    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
+	tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
+	quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
+	consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
+	cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
+	non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
+`),
+		},
+
+		"With invalid JSON object": {
+			value: []byte("{[,]}"),
+		},
```

Comment:
Same comment about spacing between test cases and the unnecessary split in the extended form.

Test cases should be scoped: keep together the ones that have a similar output. In this test, for example, we expect them all to return errors in the end, so there's no need to add a blank line between the test cases. If we had success and error ones, then I agree on separating the success block from the error block (that's what we do in all of our tests).